File: Sources/AltiveChatCore/ChatDraftPolicy.swift
import Foundation

/// 入力長を数える単位。
public enum ChatDraftLengthUnit: Hashable, Sendable {
  /// Swift の `Character` 単位。
  case characters

  /// バックエンド契約で利用される UTF-16 code unit 単位。
  case utf16
}

/// チャット入力の正規化、文字数計測、上限制御をまとめる方針。
public struct ChatDraftPolicy: Hashable, Sendable {
  /// 入力可能な最大長。`nil` は無制限。
  public let maximumLength: Int?

  /// 文字数表示を開始する長さ。`nil` は文字数を表示しない。
  public let warningThreshold: Int?

  /// 入力長を数える単位。
  public let lengthUnit: ChatDraftLengthUnit

  /// 入力方針を作成する。
  public init(
    maximumLength: Int?,
    warningThreshold: Int? = nil,
    lengthUnit: ChatDraftLengthUnit = .characters
  ) {
    let normalizedMaximum = maximumLength.map { max(0, $0) }
    self.maximumLength = normalizedMaximum
    self.warningThreshold = warningThreshold.map { threshold in
      min(max(0, threshold), normalizedMaximum ?? threshold)
    }
    self.lengthUnit = lengthUnit
  }

  /// 上限も文字数表示も持たない方針。
  public static let unrestricted = ChatDraftPolicy(maximumLength: nil)

  /// 入力値の長さを方針に沿って返す。
  public func length(of value: String) -> Int {
    switch lengthUnit {
    case .characters:
      value.count
    case .utf16:
      value.utf16.count
    }
  }

  /// 入力値を上限以内の完全な `Character` 列へ丸める。
  public func limited(_ value: String) -> String {
    guard let maximumLength, length(of: value) > maximumLength else { return value }
    switch lengthUnit {
    case .characters:
      return String(value.prefix(maximumLength))
    case .utf16:
      var currentLength = 0
      return String(
        value.prefix { character in
          let characterLength = String(character).utf16.count
          guard currentLength + characterLength <= maximumLength else { return false }
          currentLength += characterLength
          return true
        }
      )
    }
  }

  /// 前後の空白と改行を除き、空文字の場合は `nil` を返す。
  public func normalizedText(from value: String) -> String? {
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalized.isEmpty else { return nil }
    guard maximumLength.map({ length(of: normalized) <= $0 }) ?? true else { return nil }
    return normalized
  }

  /// 現在の長さを表示すべきか返す。
  public func shouldShowLength(for value: String) -> Bool {
    guard let warningThreshold else { return false }
    return length(of: value) >= warningThreshold
  }
}

File: lib/src/models/chat_draft_policy.dart
import 'package:characters/characters.dart';
import 'package:flutter/foundation.dart';

/// 入力長を数える単位。
enum ChatDraftLengthUnit {
  /// Unicodeの拡張書記素クラスタ単位。
  characters,

  /// バックエンド契約で利用されるUTF-16 code unit単位。
  utf16,
}

/// チャット入力の正規化、文字数計測、上限制御をまとめる方針。
@immutable
class ChatDraftPolicy {
  /// 入力方針を作成する。
  const ChatDraftPolicy({
    int? maximumLength,
    int? warningThreshold,
    this.lengthUnit = ChatDraftLengthUnit.characters,
  }) : maximumLength = maximumLength == null
           ? null
           : maximumLength < 0
           ? 0
           : maximumLength,
       warningThreshold = warningThreshold == null
           ? null
           : warningThreshold < 0
           ? 0
           : maximumLength == null
           ? warningThreshold
           : warningThreshold > (maximumLength < 0 ? 0 : maximumLength)
           ? (maximumLength < 0 ? 0 : maximumLength)
           : warningThreshold;

  /// 上限も文字数表示も持たない方針。
  const ChatDraftPolicy.unrestricted()
    : maximumLength = null,
      warningThreshold = null,
      lengthUnit = ChatDraftLengthUnit.characters;

  /// 入力可能な最大長。`null`は無制限。
  final int? maximumLength;

  /// 文字数表示を開始する長さ。`null`は文字数を表示しない。
  final int? warningThreshold;

  /// 入力長を数える単位。
  final ChatDraftLengthUnit lengthUnit;

  /// 入力値の長さを方針に沿って返す。
  int length(String value) {
    return switch (lengthUnit) {
      ChatDraftLengthUnit.characters => value.characters.length,
      ChatDraftLengthUnit.utf16 => value.codeUnits.length,
    };
  }

  /// 入力値を上限以内の完全なCharacter列へ丸める。
  String limited(String value) {
    final maximumLength = this.maximumLength;
    if (maximumLength == null || length(value) <= maximumLength) {
      return value;
    }

    final characters = value.characters;
    switch (lengthUnit) {
      case ChatDraftLengthUnit.characters:
        return characters.take(maximumLength).toString();
      case ChatDraftLengthUnit.utf16:
        var currentLength = 0;
        final result = StringBuffer();
        for (final character in characters) {
          final characterLength = character.codeUnits.length;
          if (currentLength + characterLength > maximumLength) {
            break;
          }
          result.write(character);
          currentLength += characterLength;
        }
        return result.toString();
    }
  }

  /// 前後の空白と改行を除き、送信可能な本文を返す。
  String? normalizedText(String value) {
    final normalized = value.trim();
    if (normalized.isEmpty) {
      return null;
    }
    if (maximumLength case final maximumLength?) {
      if (length(normalized) > maximumLength) {
        return null;
      }
    }
    return normalized;
  }

  /// 現在の長さを表示すべきか返す。
  bool shouldShowLength(String value) {
    final warningThreshold = this.warningThreshold;
    return warningThreshold != null && length(value) >= warningThreshold;
  }
}
